python/phase-2.md

```diff
@@ -993,4 +1035,192 @@ Notes:
+  parameter is mostly only used for aligning names and not for general inheritance-based code reuse
+  * The implements does enforce that the class actually does implement that
+  * Cannot provide a name if you provide `implements=` (that base one provides the name)
+* Rules for general inheritance:
+  * There can be only one `@workflow.run` (but it can be overridden)
+  * Every method override of a `@workflow.<x>` _must_ have the same decorator and decorator args and method args
+    * So overridding, say, run or a signal handler or whatever must have the same
+    * We probably can't easily confirm the arg type annotations are equivalent due to so many possibilities with
+      generics and such
+* This means that classes marked as workflows will be typed as their `implements`, _not_ their actual class type, which
+  should be ok
+
+### Dynamic Signals/Queries
+
+There is a concern that we won't allow easy runtime signal/query handler management from inside the workflow. Currently,
+we allow catch-all signal and query handlers like so:
+
+```python
+@workflow.defn
+class MyWorkflow:
+    @workflow.run
+    async def run(self):
+        ...
+
+    @workflow.signal(dynamic=True)
+    async def default_signal_handler(self, signal_name: str, *args: Any) -> None:
+        ...
+
+    @workflow.query(dynamic=True)
+    def default_query_handler(self, query_name: str, *args: Any) -> Any:
+        ...
+```
+
+So technically you could have a runtime mixin helper like so:
+
+```python
+class DynamicHandlerMixin:
+    def __init__(self, *args, **kwargs) -> None:
+        ...
+
+    def set_signal_handler(self, signal_name: str, callback: Callable) -> None:
+        ...
+
+    @workflow.signal(dynamic=True)
+    async def default_signal_handler(self, signal_name: str, *args: Any) -> None:
+        ...
+
+    def set_query_handler(self, signal_name: str, callback: Callable) -> None:
+        ...
+
+    @workflow.query(dynamic=True)
+    def default_query_handler(self, query_name: str, *args: Any) -> Any:
+        ...
+```
+
+Which could be used like:
+
+```python
+@workflow.defn
+class MyWorkflow(DynamicHandlerMixin):
+    def __init__(self) -> None:
+        ...
+
+    @workflow.run
+    async def run(self):
+        if some_conditional():
+            self.set_signal_handler("my signal name", self.some_signal_handler)
+
+    def some_signal_handler(s: str) -> None:
+        ...
+```
+
+But there are downsides, namely:
+
+* This would require user-land to buffer the signals instead of us
+* This would require user-land to throw on missing query instead of our error
+  which includes known registered names
+* You have to opt-in to this
+
+Instead, we can offer this in the `temporalio.workflow` package:
+
+```python
+def set_signal_handler(name: str, handler: Callable) -> None:
+    ...
+
+def set_query_handler(name: str, handler: Callable) -> None:
+    ...
```

Comment:
What I'm wondering is if we can do something like this (solving the req-resp problem as an example). Signals are still statically defined, but, reusable.

```python
class ReqRespSignalUser:
    def __init__(self):
        self.res = asyncio.Future()

    async def request_handler(self, arg):
        raise NotImplementedError

    async def make_request(self, external_wf_handle, arg):        
        await external_wf_handle.signal(ReqRespSignalUser.req_resp_sig, "hi") 
        return await self.res

    @workflow.signal
    async def req_resp_sig(self, req_or_resp: str) -> None:
        # Made up but you get the idea
        if req_or_resp.is_req:
            self.request_callback(req_or_resp)
        else:
            self.res.set_result(req_or_resp)

@workflow.defn
class MyWorkflow(ReqRespSignalUser):
    # Now this workflow can call self.make_request(other_workflow, "whatever")
    
    def request_handler(self, arg):
        return "request handled bruh"
```

Reply:
What about a class that is not part of the workflow? Think about a library class.

Reply:
I think this works for that - since the library author would have to know to `set_signal_handler` as the traditional way, instead they make a class like this one which explicitly defines the signal handler, and then that class can be inherited by the workflow, or by another class which is eventually inherited by the workflow, etc.

Reply:
The problem described in this section is for dynamically _named_ signals/queries. The req/resp for the _same named_ signal is next section. The solution you're showing doesn't relate to dynamically named signals.

We need `set_signal_handler` and `set_query_handler` for dynamically named signals/queries. We could use user-land code and the catch-all signal/query handler but I listed the downsides. After some thought, I figured the ability to set these handlers from inside the workflow was acceptable (though we'd encourage the other approach of course).

Reply:
What I'm getting at is that dynamic names are not really required. You can accomplish all the same things you'd want to do with them by switching on some message type, as you know and we talked about before.

So the question becomes, what do dynamic names solve for you, and being able to have self-contained logic that depends on signals is one reason and my example shows how you can still have that self contained behavior w/o a dynamically named signal.

So I'm wondering if we can come up with a motivating example that shows how dynamically named signals are actually needed to solve some problem that can't be solved this way, I can't quite think of one and that'd be a good thing to have so we can know "Oh ok we do need them for such and such reason"

Reply:
> So I'm wondering if we can come up with a motivating example that shows how dynamically named signals are actually needed to solve some problem that can't be solved this way, I can't quite think of one and that'd be a good thing to have so we can know "Oh ok we do need them for such and such reason"

I've been told "DSL"s, e.g.

```python
@dataclass
class DslInput:
  signals: Dict[str, Action]
  queries: Dict[str, TypeOfData]

@workflow.defn
class MyDslWorkflow:
  @workflow.rum
  async def run(input: DslInput):
    for name, action in input.signals
      workflow.set_signal_handler(name, action_handler_for(action))
    for name, type_of_data in input.queries
      workflow.set_query_handler(name, query_handler_for(type_of_data))
```

The reason we would support this natively instead of the user-land `dynamic=True` approach of my mixin above are in the bullets I laid out.

Reply:
Seems to me this can still easily work the same way as my example - the DSL input can be switched on, either directly in the workflow or in some kind of mix-in like my example.

We know we get some nice benefits with static only signals, and it still seems to me nothing necessitates dynamic ones, or is even really substantially easier with them.

If we want to add it to stay consistent, I suppose that's fine, but seems like a good opportunity to simplify in the absence of a super compelling reason to keep them

Reply:
> Seems to me this can still easily work the same way as my example - the DSL input can be switched on, either directly in the workflow or in some kind of mix-in like my example.

But you don't get custom names driven by the input, or if you do, you don't get the benefit of Temporal logic for not-found signal/query handlers/names.

> We know we get some nice benefits with static only signals, and it still seems to me nothing necessitates dynamic ones, or is even really substantially easier with them.

The example I gave of custom, dynamic naming while still getting all the benefits of the default Temporal logic handling of unknown signals/queries necessitates it. I think to support your argument you have to either:

* Convince everyone that you don't need dynamically named signals/queries (e.g. from workflow input)
* Show how you could do this at runtime and still get the same Temporal logic/errors of unhandled signals/queries

> If we want to add it to stay consistent, I suppose that's fine, but seems like a good opportunity to simplify in the absence of a super compelling reason to keep them

I don't want to take that opportunity here personally to make that decision. But I definitely agree that we should discourage this approach, probably going as far as not documenting its use when documenting normal signal/query documentation.

Reply:
What I can do if it's worth it is mark `set_signal_handler` and `set_query_handler` as experimental (instead of just discouraged) and if you can convince the team of removal, I can remove.